src/compiler/optimizations/utils.cc

```diff
@@ -63,7 +63,7 @@ Type compute_guaranteed_type(Expression* node, Class* holder, Method* method) {
     return field->type();
   } else if (node->is_Typecheck()) {
     auto check = node->as_Typecheck();
-    return check->type();
+    if (check->is_as_check()) return check->type();
```

Comment:
I theory we know that the type is 'bool' otherwise, but I don't see an easy way to get to that type here.

Reply:
Yeah, I thought about that. The type checker knows this (and more).